python/ray/air/_internal/remote_storage.py

```diff
@@ -173,6 +159,84 @@ def _is_local_windows_path(path: str) -> bool:
     return False
 
 
+def _translate_s3_options(options: Dict[str, List[str]]) -> Dict[str, Any]:
+    """Translate pyarrow s3 query options into s3fs storage_kwargs"""
+    # Map from s3 query keys --> s3fs kwarg names
+    option_map = {
+        "endpoint_override": "endpoint_url",
+        "region": "region_name",
+        "access_key": "aws_access_key_id",
+        "secret_key": "aws_secret_access_key",
+    }
+
+    client_kwargs = {}
+    for opt, target in option_map.items():
+        if opt in options:
+            client_kwargs[target] = options[opt][0]
+
+    # s3fs directory cache does not work correctly, so we pass
+    # `use_listings_cache` to disable it. See https://github.com/fsspec/s3fs/issues/657
+    return {"client_kwargs": client_kwargs, "use_listings_cache": False}
+
+
+def _translate_gcs_options(options: Dict[str, List[str]]) -> Dict[str, Any]:
+    """Translate pyarrow gcs query options into gcsfs storage_kwargs"""
+    # Map from gcs query keys --> gcsfs kwarg names
+    option_map = {
+        "endpoint_override": "endpoint_url",
+    }
+
+    storage_kwargs = {}
+    for opt, target in option_map.items():
+        if opt in options:
+            storage_kwargs[target] = options[opt][0]
+
+    return storage_kwargs
+
+
+def _get_fsspec_fs_and_path(uri: str) -> Optional["pyarrow.fs.FileSystem"]:
+    parsed = urllib.parse.urlparse(uri)
+
+    storage_kwargs = {}
+    if parsed.scheme in ["s3", "s3a"] and parsed.query:
+        storage_kwargs = _translate_s3_options(urllib.parse.parse_qs(parsed.query))
+    elif parsed.scheme in ["gs", "gcs"] and parsed.query:
+        storage_kwargs = _translate_gcs_options(urllib.parse.parse_qs(parsed.query))
+
+    try:
+        fsspec_fs = fsspec.filesystem(parsed.scheme, **storage_kwargs)
+    except Exception:
```

Comment:
I think we can catch a `ValueError`.

Reply:
I've added a comment to explain that in the code